dds/DCPS/RTPS/Sedp.cpp

```diff
@@ -2735,8 +2835,8 @@ Sedp::association_complete(const RepoId& localId,
     write_durable_dcps_participant_secure(remoteId);
   } else if (remoteId.entityId == ENTITYID_P2P_BUILTIN_PARTICIPANT_VOLATILE_SECURE_READER) {
 
-    if (associated_volatile_readers_.insert(remoteId).second) {
-
+    if (associated_volatile_readers_.insert(remoteId).second && spdp_.remote_is_requester(remoteId)) {
+      // TODO: Will reliability resend these?
```

Comment:
(Not directly related to the "TODO" comment above)

Take a look at the order of `Sedp::associate_volatile` (which gets the SharedSecret from Auth to Crypto and makes the transport aware of it) relative to sending the Final part of the handshake in `Spdp::handle_handshake_message`.

In order to know that key exchange on the "volatile" built-in endpoints will be able to decrypt messages, the installation of the SharedSecret on the "requester" side should occur before sending the Final handshake message.

On the "replier" side, keys can be sent directly after receiving the Final and installing the SharedSecret.

On the "requester" side, keys can be sent after receiving at least one key from the peer.  To guarantee this is possible, the participant key could be changed from conditionally to unconditionally generated.  Still need to check how the spec handles this, what it can/should guarantee.